Добавлен фильтр для премиум пользователей в UserService

MaksimMas opened the pull request:
Приоритет поиска для премиум-пользователей

MaksimMas committed:
Добавлен фильтр для премиум пользователей в UserService

Timka-cloud1 commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,29 @@
+                .flatMap(filter -> filter.apply(userRepository.findPremiumUsers(),userFilterDto))
+                .map(userMapper::toDto)
+                .toList();
+    }
```
если я правильно понимаю то ты каждый раз кидаешь запрос в БД для каждого фильтра. Это ломает логику фильтра и лишняя нагрузка.

Timka-cloud1 commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -0,0 +1,23 @@
+        if (userFilterDto == null){
+            throw new NullUserFilterException("Empty filter");
+        }
+        return userService.getPremiumUsers(userFilterDto);
```
может если фильтр null то стоит просто отдавать всех премиум пользователей без фильтров?

MaksimMas committed:
Исправил логику фильтрации

AlexeyAntipin commented on src/main/java/school/faang/user_service/filters/user/UserCityPattern.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.stream.Stream;
+
+@Component
+public class UserCityPattern implements UserFilter{
```
Перед открывающейся скобкой должен быть пробел)

AlexeyAntipin commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -0,0 +1,23 @@
+public class UserController {
+    private final UserService userService;
+
+    public List<UserDto> getPremiumUsers(UserFilterDto userFilterDto){
```
Перед открывающейся скобкой должен быть пробел)

AlexeyAntipin commented on src/main/java/school/faang/user_service/filters/user/UserNamePattern.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.stream.Stream;
+
+@Component
+public class UserNamePattern implements UserFilter{
```
Перед открывающейся скобкой должен быть пробел)

AlexeyAntipin commented on src/main/java/school/faang/user_service/filters/user/UserNamePattern.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public List<User> apply(List<User> users, UserFilterDto filter){
+        return users.stream().filter(user -> user.getUsername().contains(filter.getNamePattern())).toList();
```
Судя по названию, namePattern подразумевает именно шаблон для userName, то есть использование регулярных выражений. В таком случае лучше использовать не contains, а matches

AlexeyAntipin commented on src/main/java/school/faang/user_service/service/skills/SkillService.java:
```diff
@@ -1,17 +1,88 @@
+    }
+
+    private Skill findSkillById(long skillId) {
+        return skillRepository.findById(skillId).orElseThrow(() -> new EntityNotFoundException("Skill not found by id " + skillId));
```
Это не совсем бизнес логика. По сути ты здесь просто делаешь апи к репозиторию чуть более удобным. Предлагаю создать в пакете repository пакет adapter и там создать SkillRepositoryAdapter, после чего этот метод перенести туда. Таким образом мы сможем более корректно отделить слои друг от друга, а в бизнес логике реализовывать именно бизнес логику, а не логику слоя-адаптера

MaksimMas replied:
Это не мой код, я объединил два SkillService в один, поскольку у нас в проекте создавалось два бина и они конфликтовали между собой

AlexeyAntipin commented on src/test/java/school/faang/user_service/service/user/UserServiceTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    @BeforeEach
+    void init(){
+        mockUserRepo = Mockito.mock(UserRepository.class);
```
Можно добавить на класс аннотацию @ExtendsWith(MockitoExtensions.class), после чего над mockUserRepo повесить аннотацию @Mock. Таким образом, код теста станет короче. Да и в целом, annotation based реализации считается более предпочтительной
Так можно сделать со всеми mock-классами

MaksimMas committed:
Исправил тесты
Поменял UserService, теперь если фильтров нет пользователю возвращается не ошибка, а просто все premiumUsers

AlexeyAntipin commented:
Нужно избавиться от конфликтов

MaksimMas committed:
Merge branch 'hydra-master-stream8' into hydra-stream8-BJS2-52679
